### PR TITLE
[Parse] Implementation for SE-200 (raw strings)

### LINKED ISSUE #48912: [SR-6362] Raw escaping in string literals
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-6362      |
   |Radar             | rdar://problem/47725014         |
   |Original Reporter | @johnno1962      |
   |Type              | New Feature    |
   |Status            | Closed        |
   |Resolution        | Done    |
        
   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | New Feature, LanguageFeatureRequest        |
   |Assignee          | @johnno1962      |
   |Priority          | Medium      |

   

   md5: 0d6a3785983c156c861de08a567e4986

  </details>

**Issue Description:**

Often it is inconvenient and opaque entering string literals that are in tended as regex patterns into a swift source.

Taking the lead from python it perhaps a new syntax `r"\w+\s+\w"` could be added to the swift language. This would be equivalent to `" w+ s+ w"`. The specification required would be that if a string literal is preceded by r then unknown escapes are passed though to the literal as-is including the back-slash character rather than going an error. This would be a very minor change confined to Lexer.cpp that I'll work on tomorrow and present to swift-evolution.

### PR DESCRIPTION
**Pull request by johnno1962**
As per discussion in second pitch of raw strings on swift-evolution https://forums.swift.org/t/pure-bikeshedding-raw-strings-why-yes-again/13866. Work in progress, submitting now to get PR link.

Resolves #48912 and also lays groundwork for #51192.

### CODE REVIEW

**Comment by johnno1962** [comment]
@gregomni, any chance you could do a @swift-ci please test for me so I can see what this breaks?

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
(just doing a drive by @swift-ci)

**Comment by johnno1962** [comment]
Cheers.

File: test/Parse/raw_string.swift [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+// CHECK: "Alice: \"How long is forever?\" White Rabbit: \"Sometimes, just one second.\""
+
+_ = #"\#\#1"#
+/// CHECK: "\\#1"
+
+_ = ##"\#1"##
+/// CHECK: "\\#1"
+
+_ = #"c:\windows\system32"#
+/// CHCECK: "c:\\windows\\system32"
```
Comment (benrimmington) [inline test/Parse/raw_string.swift]: Typo: `CHCECK` => `CHECK`

Does it matter how many forward slash characters are used?
i.e. `/// CHECK:` versus `// CHECK:`

Would the following also be a valid raw string test?

```
_ = #"""
###################################################################
## This source file is part of the Swift.org open source project ##
###################################################################
"""#
```

**Comment by johnno1962** [comment]
Tests fixed up and I’ve added the one you suggest and tested.

File: test/Parse/raw_string.swift [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+_ = #"""
+    a string with
+    """
+    in it
+    """#
+// CHECK: "a string with\n\"\"\"\nin it"
+
+_ = #"a raw string containing \r\n"#
+// CHECK "a raw string containing \\r\\n"
```
Comment (benrimmington) [inline test/Parse/raw_string.swift]: Missing colon. `// CHECK` => `// CHECK:`

Comment (johnno1962) [reply test/Parse/raw_string.swift]: Crikey. Ammended.

File: test/Parse/raw_string.swift [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+import Swift
+
+_ = #"""
+###################################################################
+## This source file is part of the Swift.org open source project ##
+###################################################################
+"""#
+
+// CHECK: "###################################################################\n## This source file is part of the Swift.org open source project ##\n###################################################################"
```
Comment (benrimmington) [inline test/Parse/raw_string.swift]: You might be able to use regular expressions in the `CHECK:` directive.

```swift
// CHECK: "{{#{67}}}\n## This source file is part of the Swift.org open source project ##\n{{#{67}}}"
```

<https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-pattern-matching-syntax>

Or shorter raw strings which test the same idea. For example, using Markdown headers.

```swift
_ = #"""
    # H1 #
    ## H2 ##
    ### H3 ###
    """#
// CHECK: "# H1 #\n## H2 ##\n### H3 ###"
```

Do any of the tests require the `--strict-whitespace` option?

<https://llvm.org/docs/CommandGuide/FileCheck.html#cmdoption-strict-whitespace>

Comment (johnno1962) [reply test/Parse/raw_string.swift]: Couldn’t get the regex to work but I added your extra test and --strict-whitespace. Thanks

File: include/swift/Parse/Token.h [hunk 4]
```diff
@@ -266,17 +269,23 @@ class Token {
...
   void setToken(tok K, StringRef T, unsigned CommentLength = 0,
-                bool MultilineString = false) {
+                bool MultilineString = false, unsigned DelimiterLength = 0) {
     Kind = K;
     Text = T;
     this->CommentLength = CommentLength;
     EscapedIdentifier = false;
     this->MultilineString = MultilineString;
+    StringDelimiterLength = DelimiterLength;
+    assert(StringDelimiterLength == DelimiterLength && "delimiter too long");
```
Comment (rintaro) [inline include/swift/Parse/Token.h]: This `assert` never hits.

File: lib/Parse/Lexer.cpp [hunk 5]
```diff
@@ -1645,13 +1707,14 @@ static void validateMultilineIndents(const Token &Str,
 /// lexStringLiteral:
 ///   string_literal ::= ["]([^"\\\n\r]|character_escape)*["]
 ///   string_literal ::= ["]["]["].*["]["]["] - approximately
-void Lexer::lexStringLiteral() {
+void Lexer::lexStringLiteral(unsigned DelimiterLength) {
   const char *TokStart = CurPtr-1;
   assert((*TokStart == '"' || *TokStart == '\'') && "Unexpected start");
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: Please adjust `TokStart` with `DelimiterLength` as soon as possible. Otherwise, for unterminated string such as `#"foo<EOF>`, `#` is not included in the unknown token.

File: lib/Parse/Lexer.cpp [hunk 6]
```diff
@@ -1660,12 +1723,16 @@ void Lexer::lexStringLiteral() {
     if (*CurPtr != '\n' && *CurPtr != '\r')
       diagnose(CurPtr, diag::lex_illegal_multiline_string_start)
         .fixItInsert(Lexer::getSourceLoc(CurPtr), "\n");
+    ExtraTermination.insert(ExtraTermination.size(), 2, *TokStart);
   }
+  ExtraTermination.insert(ExtraTermination.size(), DelimiterLength, '#');
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: You can use `append()` here.

File: lib/Parse/Lexer.cpp [hunk 7]
```diff
@@ -1645,13 +1707,14 @@ static void validateMultilineIndents(const Token &Str,
...
 ///   string_literal ::= ["]["]["].*["]["]["] - approximately
-void Lexer::lexStringLiteral() {
+void Lexer::lexStringLiteral(unsigned DelimiterLength) {
   const char *TokStart = CurPtr-1;
   assert((*TokStart == '"' || *TokStart == '\'') && "Unexpected start");
   // NOTE: We only allow single-quote string literals so we can emit useful
   // diagnostics about changing them to double quotes.
 
   bool wasErroneous = false, MultilineString = false;
+  std::string ExtraTermination;
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: `SmallString<8>` is sufficient I think.

File: lib/Parse/Lexer.cpp [hunk 8]
```diff
@@ -1731,20 +1797,19 @@ void Lexer::lexStringLiteral() {
...
-          CurPtr += 2;
-          formToken(tok::string_literal, TokStart, MultilineString);
-          if (Diags)
-            validateMultilineIndents(NextToken, Diags);
-          return;
-        }
-        else
-          continue;
+      // Is this the end of multiline/delimited string literal?
+      if (StringRef(CurPtr, ExtraTermination.length()) == ExtraTermination) {
```
[the remark this answers is not in the document]
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: Hi @rintaro, I was going to ask you for a review, particularly for the changes to skipToEndOfInterpolatedExpression()

File: include/swift/Parse/Token.h (hunk 4 above)
Comment (johnno1962) [reply include/swift/Parse/Token.h]: I does, honest. StringDelimiterLength is only 8 bits

File: lib/Parse/Lexer.cpp (hunk 6 above)
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: Thanks, fixed

File: lib/Parse/Lexer.cpp (hunk 5 above)
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: Difficult to know what to do without broader changes to lexStringLiteral(). This is fixed though.

File: lib/Parse/Lexer.cpp (hunk 7 above)
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: Thanks, fixed.

**Comment by rintaro** [comment]
I'll review this closely today.

File: lib/Parse/Lexer.cpp [hunk 9]
```diff
@@ -1262,14 +1295,26 @@ unsigned Lexer::lexCharacter(const char *&CurPtr, char StopQuote,
       diagnose(CurPtr-1, diag::lex_unterminated_string);
     return ~1U;
   case '\\':  // Escapes.
+    if (!delimiterMatches(DelimiterLength, CurPtr))
+      return '\\';
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: I don't think this is right. For `###" \#(some) "###`, this function should diagnose it as invalid escape instead of just returning `\`.

File: lib/Parse/Lexer.cpp [hunk 10]
```diff
@@ -1731,20 +1799,19 @@ void Lexer::lexStringLiteral() {
...
-          CurPtr += 2;
-          formToken(tok::string_literal, TokStart, MultilineString);
-          if (Diags)
-            validateMultilineIndents(NextToken, Diags);
-          return;
-        }
-        else
-          continue;
+      // Is this the end of multiline/delimited string literal?
+      if (StringRef(CurPtr, BufferEnd - CurPtr).startswith(ExtraTermination)) {
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: What about `#" "##" "#`? is this two `#" "#`s or one literal containing `' "##" '`?

**Review by rintaro — CHANGES_REQUESTED** [review]
Overall looks great! Please add test cases for various invalid cases, including new diagnostics you added.

File: lib/Parse/Lexer.cpp (hunk 10 above)
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: Two literals I think though others have questioned this. “# is the termination unless we make a rule it can’t be followed by a #

File: lib/Parse/Lexer.cpp [hunk 11]
```diff
@@ -1204,6 +1206,36 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
...
+/// match the number of # characters after \ inside the string? This allows
+/// interpolation inside a "raw" string. Normal/cooked string processing is
+/// the degenerate case of there being no # characters surrounding the quotes.
+/// If delimiter matches, advances byte pointer passed in and returns true.
+static bool delimiterMatches(unsigned DelimiterLength, const char *&BytesPtr) {
+  if (!DelimiterLength)
+    return true;
+  for (unsigned i = 0; i < DelimiterLength; i++)
+    if (BytesPtr[i] != '#')
+      return false;
```
[the remark this answers is not in the document]
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: This is checked for in lexCharacter() where it can give a diagnostic.

File: lib/Parse/Lexer.cpp (hunk 10 above)
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: I’ve added the rule that a raw string can’t be terminated with a # immediately after it which people seem to expect for strings such as `#" "##" “#` I’ll ask the forum if this makes sense.

File: lib/Parse/Lexer.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (rintaro) [reply lib/Parse/Lexer.cpp]: Ah, my bad. You are absolutely right. Sorry for the noise!

**Comment by johnno1962** [comment]
Toolchain available for testing 
http://johnholdsworth.com/swift-RAWSTRINGII-DEVELOPMENT-SNAPSHOT-2018-06-27-a.tgz

File: lib/Parse/Lexer.cpp (hunk 10 above)
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: Evolution didn’t seem to keen on the idea just when I’ve decided this was the best way to go.

**Comment by johnno1962** [comment]
Hi @xwu, perhaps it’s time for you to have a look at this code for “raw strings” which is essentially ready. The last commit is my take on checking for zero-width characters which can be backed out i people want to take a different road. I'd create a separate PR but I expect SE-200 to go through and want to get out of the loop. This implementation contains the “no-# after delimiter” idea though it can also be backed out.

**Review by xwu — COMMENTED** [review]
This PR was a little challenging to review because it wasn't squashed, yet you specifically called out the last commit as separable from the rest.

So I went through one commit at a time. A lot of comments for the first commit and many fewer for subsequent ones--it was enjoyable to see the refinement process. That said, it means that some comments are completely inapplicable and others still apply but are hidden because the code has changed in unrelated ways in subsequent commits. Please take a look through those hidden comments.

In general, I really like the implementation. It has very much improved and seems quite mature.

However, the last commit is far from adequate to address the issue of invisible characters; this really ought to be separated out entirely. It's also a mightily long topic very much deeper than can be covered here. There is a languishing proposal draft of thousands of words to deal with this type of security issue.

File: lib/Parse/Lexer.cpp [hunk 12]
```diff
@@ -1206,6 +1206,38 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
   }
 }
 
+/// diagnoseZeroWidth - check for and error zero-width characters in delimiters
+static bool diagnoseZeroWidth(const char *&CurPtr, DiagnosticEngine *Diags) {
+  const unsigned char *TmpPtr = (const unsigned char *)CurPtr;
+  // zero-width set assumed to be: U+200B, U+200C, U+200D, U+2060, U+FEFF
```
[the remark this answers is not in the document]
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: Thanks for your comments. I’ll work through them tomorrow. I looked at supporting ignoring invisible characters rather than reporting an error and it would be _much_ harder to implement as the relationship between the token span and the actual literal will become +/- a variable length rather than just (1 or 3) + DelimiterLength. Besides I don’t think it's a good idea - more prone to bugs though I see where you are coming from with other whitespace.

I figured there would be a difficult to anticipate set of zero-width chars to exclude which is going to make this difficult to make watertight. I think this is an implementation detail of diagnoseZeroWidth() which I’d like to leave in this PR in its incomplete but architected form while we’re working in the neighbourhood as a start for more comprehensive treatment by someone else later. Not sure how to use ICU but doesn’t that requite 16 bit chars? So I’d like to not revert the last commit if that’s ok before layering the next commit on top of it.

[the remark this answers is not in the document]
Comment (xwu) [reply lib/Parse/Lexer.cpp]: Personally, I feel _very strongly_ against leaving this commit in. It's not addressing the issue in any systematic way but hides the easiest-to-demonstrate manifestation of it.

Either we believe that this is not a major issue or we believe that it is. In the former case, we can fix this for all multi-character string delimiters in some systematic way at a later time. In the latter case, we _must_ fix this for all multi-character string delimiters in some systematic way as soon as possible. But in either case, I can see no justification for a partial implementation now, which has the problem of hiding the issue but not solving it.

**Review by johnno1962 — COMMENTED** [review]
I couldn’t bring myself to revert the zero-width characters commit, sorry, there was too much useful groundwork for an eventual solution. I did however comment out a new but still limited implementation of diagnoseZeroWidth() so it does not mask that a fulsome solution has not currently been found (if it is even possible - I couldn’t find anything in ICU that allowed you to check against the class of zero-width characters.)  I’ve also backed out the don’t terminate if # after closing delimiter idea which found no traction. If you could make a final pass over the full set of file changes we should be good to go.

File: lib/Parse/Lexer.cpp [hunk 13]
```diff
@@ -2281,16 +2310,9 @@ void Lexer::lexImpl() {
...
-      const char *Lookahead = CurPtr;
-      while (*Lookahead == '#')
-        Lookahead++;
-      if (*Lookahead++ == '"') {
-        unsigned DelimiterLength = Lookahead - CurPtr;
-        CurPtr = Lookahead;
-        return lexStringLiteral(DelimiterLength);
-      }
-    }
+  case '#':
```
[the remark this answers is not in the document]
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: I’m just following other code thereabouts.

**Comment by johnno1962** [comment]
I’ve committed another attempt at this problem using a static table generated using code that looks for non-changes in the display size.width of an attributed string “\”%C\””. There are 612 unichars found. There is still the problem of 20 bit unicode values :( some of which will be zero-width no doubt.

**Comment by xwu** [comment]
@johnno1962 A valiant effort, but I too doubt that this is the right approach. For the moment, if you're going to back out the idea about changing the parsing of the closing delimiter, then this is entirely a non-issue for this PR, because any such characters would cause the remaining text to be flagged as invalid syntax.

On a cursory view, the remainder of the revisions look fine, but it'll be a few days before I have time to read through the whole thing.

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
Looks spurious.

**Comment by brentdax** [comment]
@swift-ci please test

File: lib/Parse/Lexer.cpp [hunk 14]
```diff
@@ -1204,6 +1206,202 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
   }
 }
 
+const static uint32_t ZeroWidthC[] = {
```
Comment (brentdax) [inline lib/Parse/Lexer.cpp]: Should we extract a function to test for zero-width characters and put it in swift/Basic/Unicode.{h,cpp}?

I can't speak for the accuracy of this list of characters, but a lot of them seem to be visible but take up no extra space—for instance, I pulled out U+1A59 at random and it turned out to be a combining mark. Do we want to diagnose those kinds of characters?

File: lib/Parse/Lexer.cpp [hunk 15]
```diff
@@ -1204,6 +1206,202 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
...
+  //  the display width of attributed strings checking when it does not change.
+  //  The font used was SF Mono, 11pt, the default font of the Xcode editor.
+  static std::vector<uint32_t> ZeroWidthV;
+  if (!ZeroWidthV.size())
+    ZeroWidthV.assign(ZeroWidthC,
+                      ZeroWidthC + sizeof ZeroWidthC/sizeof ZeroWidthC[0]);
+
+  const char *TmpPtr = CurPtr;
+  while (true) {
+    uint32_t NextChar = validateUTF8CharacterAndAdvance(TmpPtr, TmpPtr + 6);
```
Comment (brentdax) [inline lib/Parse/Lexer.cpp]: `TmpPtr + 6` could be past the end of the buffer, right? Are we certain that even a malicious file could not cause `validateUTF8CharacterAndAdvance()` to read past the end?

**Comment by brentdax** [comment]
@johnno1962 Looks like some of the error messages in test/Parse/raw_string_errors.swift don't quite match.

File: lib/Parse/Lexer.cpp (hunk 14 above)
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: @xwu don’t think it shouldn't be included until we come up with a comprehensive solution for all of Swift so we’re probably going to have to back it out anyway. I say leave it in for now while we wait.

File: lib/Parse/Lexer.cpp (hunk 15 above)
Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: It shouldn't read of the buffer as the buffer always ends with a 0 which will stop validateUTF8CharacterAndAdvance processing. Anyways this will likely get backed out.

**Comment by johnno1962** [comment]
My bad there due to capitalisation and a last minute change. The error messages should be fixed.

File: lib/Parse/Lexer.cpp (hunk 15 above)
Comment (brentdax) [reply lib/Parse/Lexer.cpp]: Okay. Let me know when you've removed whatever you want to remove.

Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: OK, OK, I’ve removed the code for now.

**Comment by johnno1962** [comment]
Ready when you are @brentdax, can you run another test please?

File: lib/Parse/Lexer.cpp [hunk 16]
```diff
@@ -1204,6 +1206,68 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
   }
 }
 
+/// diagnoseZeroWidthMatchAndAdvance - Error zerowidth characters in delimiters.
+/// A non visible character in the middle of a delimter can be used to extend
+/// the literal beyond what it would appear creating potential security bugs.
+static bool diagnoseZeroWidthMatchAndAdvance(char Target, const char *&CurPtr,
+                                             DiagnosticEngine *Diags) {
+  // Detect, diagnose and skip over zero-width characters here if required.
```
Comment (xwu) [inline lib/Parse/Lexer.cpp]: Marking this as “TODO:” might be a good idea.

File: lib/Parse/Lexer.cpp [hunk 17]
```diff
@@ -1204,6 +1206,68 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
...
+  while (diagnoseZeroWidthMatchAndAdvance('#', TmpPtr, Diags))
+    CustomDelimiterLen++;
+  if (diagnoseZeroWidthMatchAndAdvance('"', TmpPtr, Diags)) {
+    CurPtr = TmpPtr;
+    return CustomDelimiterLen;
+  }
+  return 0;
+}
+
+/// delimiterMatches - Does custom delimiter (# characters surrounding quotes)
```
Comment (xwu) [inline lib/Parse/Lexer.cpp]: If you surround “#” with single quotes within this comment, it might alleviate the misread that the custom delimiter is the “number of characters surrounding quotes” (as opposed to the number of ‘#’ characters surrounding—specifically, on either side of—the quotation marks).

File: lib/Parse/Lexer.cpp [hunk 18]
```diff
@@ -2009,13 +2091,35 @@ StringRef Lexer::getEncodedStringSegment(StringRef Bytes,
...
   // range check subscripting on the StringRef.
   const char *BytesPtr = Bytes.begin();
+
+  // Special case when being called from EncodedDiagnosticMessage(...).
+  // This allows multiline and delimited strings to work in attributes.
+  // The string has already been validated by the initial parse.
+  if (IndentToStrip == ~0u && CustomDelimiterLen == ~0u) {
+    IndentToStrip = CustomDelimiterLen = 0;
+
+    // restore trailing indent removal for multiline
```
Comment (xwu) [inline lib/Parse/Lexer.cpp]: Nit: capitalize and punctuate, here and below.

File: lib/Parse/Lexer.cpp [hunk 19]
```diff
@@ -1204,6 +1206,68 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
   }
 }
 
+/// diagnoseZeroWidthMatchAndAdvance - Error zerowidth characters in delimiters.
+/// A non visible character in the middle of a delimter can be used to extend
```
Comment (xwu) [inline lib/Parse/Lexer.cpp]: Nits: “zero-width” instead of “zerowidth,” “invisible” instead of “non visible.”

Typo in the word “delimiter.”

**Review by xwu — COMMENTED** [review]
Some minor comments about comments; nothing major and can be addressed after the fact. Bravo on a great feature and implementation.

**Comment by brentdax** [comment]
@johnno1962 Yeah, I saw the same failures on the string interpolation PR.

**Comment by brentdax** [comment]
@swift-ci please smoke test and merge

**Comment by brentdax** [comment]
@swift-ci please smoke test and merge

**Comment by brentdax** [comment]
Ugh, forgot to squash it. I don't think we can clean up history on this repo once it's committed, though.

I guess I've learned a lesson about "smoke test and merge".

**Comment by johnno1962** [comment]
Thanks Brent. My fault for not squashing it myself but I wanted to keep the history. It’s going to make cherrypicking a bit of a pain though. Does anybody have permission to do a rebase -i on apple/swift to clean it up?

**Comment by johnno1962** [comment]
Could you back it out and I can file the PR again and re-merge? There are a lot of commits.

**Comment by brentdax** [comment]
@johnno1962 Would a revert without actually changing the history meet your goal? I guess if you're just trying to make sure there's one commit to cherry-pick, it would.

**Comment by brentdax** [comment]
@johnno1962 Okay, revert is in. Want to resubmit the PR?

**Comment by brentdax** [comment]
If you revert the revert commit, that should get you the right diff. From there, just fill in the commit message you want.

**Comment by johnno1962** [comment]
Top tip, thanks. I’ve created a new single commit PR. https://github.com/apple/swift/pull/19170

File: lib/Parse/Lexer.cpp [hunk 20]
```diff
@@ -1204,6 +1206,69 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
   }
 }
 
+/// diagnoseZeroWidthMatchAndAdvance - Error invisible characters in delimiters.
+/// An invisible character in the middle of a delimiter can be used to extend
+/// the literal beyond what it would appear creating potential security bugs.
+static bool diagnoseZeroWidthMatchAndAdvance(char Target, const char *&CurPtr,
+                                             DiagnosticEngine *Diags) {
+  // TODO: Detect, diagnose and skip over zero-width characters if required.
+  // See https://github.com/apple/swift/pull/17668 for possible implementation.
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: @johnno1962 This URL (this PR) doesn't have possible implementation anymore.
Could you file a JIRA issue in https://bugs.swift.org/ . And update the comment?

Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: What do you want me to do here? There is already a Jira [SR-8678](
https://bugs.swift.org/browse/SR-8678) for the zero-width problem. Should I raise a new PR to replace the link to the PR with a link to the Jira? The implementation is under this commit:
https://github.com/apple/swift/pull/17668/commits/f0f08e1e86d0c24921fe8ca71901a5ef6bd39811

Comment (rintaro) [reply lib/Parse/Lexer.cpp]: Yeah, please 1) add a link to f0f08e1 in SR-8678, and 2) raise a PR to replace the link.

Comment (johnno1962) [reply lib/Parse/Lexer.cpp]: Done: https://github.com/apple/swift/pull/19364.
